### PR TITLE
sub schema (SchemaAttributePathInstance) domain model extension + test re-factoring

### PR DESCRIPTION
**Pull request by boferri**
- incl. new init_internal_schema.sql script (+ generation via code)

### CODE REVIEW

**Comment by boferri** [comment]
test this please!

File: controller/src/docs/ui [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 22c7a9672cd998dc90fedaf9bccfa841cb6c0e77
```
Comment (janpolowinski) [inline controller/src/docs/ui:1]: please check again, is this submodule related commit intended?
(sometimes my IDE suggests me to commit changes to ../docs/ui although nothing actually changed, so I hope I didn't accidentally push this)

Comment (boferri) [inline controller/src/docs/ui:1]: @janpolowinski I think that the sub modules should be linked correctly right now in the master repo. maybe @knutwalker or @chrode can have a look at the issue here ;)

File: controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java [hunk 2]
```diff
@@ -381,7 +405,7 @@ protected Schema prepareObjectForUpdate(final Schema objectFromJSON, final Schem
 
 		super.prepareObjectForUpdate(objectFromJSON, object);
 
-		final List<AttributePath> attributePaths = objectFromJSON.getAttributePaths();
+		final Collection<SchemaAttributePathInstance> attributePaths = objectFromJSON.getAttributePaths();
```
Comment (janpolowinski) [inline controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java]: I guess the List was changed to a collection here (order not guaranteed), since we expected SchemaAPInstances to have a field order. Looking at the current SchemaAttributePathInstance impl I noticed that I skipped this for now (because I wanted to start small), so we will probably want to add a field for order like in the mapping attribute path instance. However, shouldn't we rather use a list to store the order instead of reconstructing it each time we need it from the order values?

Comment (boferri) [inline controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java]: The order of SAPI will be guaranteed by the Schema POJO itself. So I would tend to say that there is no need to change the API (method signature) there.

File: controller/src/main/java/org/dswarm/controller/resources/schema/utils/SchemaAttributePathInstancesResourceUtils.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+	@Inject
+	public SchemaAttributePathInstancesResourceUtils(final Provider<SchemaAttributePathInstanceService> persistenceServiceProviderArg,
+			final Provider<ObjectMapper> objectMapperProviderArg, final ResourceUtilsFactory utilsFactory) {
+
+		super(SchemaAttributePathInstance.class, persistenceServiceProviderArg, objectMapperProviderArg, utilsFactory);
+	}
+
+	@Override
+	public JsonNode replaceRelevantDummyIds(final SchemaAttributePathInstance object, final JsonNode jsonNode, final Set<Long> dummyIdCandidates)
```
Comment (janpolowinski) [inline controller/src/main/java/org/dswarm/controller/resources/schema/utils/SchemaAttributePathInstancesResourceUtils.java]: It would be helpful to have a short explanation / give some reasons for the replaceRelevantDummyIds method (in the super-class BasicResourceUtils)

File: controller/src/test/java/org/dswarm/controller/adapt/ControllerModelTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import org.dswarm.converter.adapt.ModelTest;
+
+/**
+ * @author tgaengler
+ */
+public class ControllerModelTest extends ModelTest {
+
+	private static final Logger log = LoggerFactory.getLogger(ControllerModelTest.class);
+
+	//@Test
```
Comment (janpolowinski) [inline controller/src/test/java/org/dswarm/controller/adapt/ControllerModelTest.java]: maybe some notes on how to use this (when to uncomment) - or is this already happening in the wiki?

File: controller/src/test/java/org/dswarm/controller/resources/job/test/TasksCsvResourceTest.java [hunk 5]
```diff
@@ -59,62 +50,37 @@
...
-import org.dswarm.persistence.model.schema.AttributePath;
-import org.dswarm.persistence.model.schema.Clasz;
 import org.dswarm.persistence.model.schema.Schema;
-import org.dswarm.persistence.service.internal.test.utils.InternalGDMGraphServiceTestUtils;
 import org.dswarm.persistence.util.DMPPersistenceUtil;
 
 public class TasksCsvResourceTest extends ResourceTest {
 
-	private static final Logger						LOG				= LoggerFactory.getLogger(TasksCsvResourceTest.class);
+	private static final Logger LOG = LoggerFactory.getLogger(TasksCsvResourceTest.class);
```
Comment (janpolowinski) [inline controller/src/test/java/org/dswarm/controller/resources/job/test/TasksCsvResourceTest.java]: Note (more for myself, will check on Monday): There are a lot of commits concerning only formatting. Is this formatting inline with our latest formatting rules, or do we need to run the formatting? (I think when I last tried the variant with tabs was still used (though I don't like it ;) .

File: controller/src/test/java/org/dswarm/controller/resources/schema/test/SchemasResourceTest.java [hunk 6]
```diff
@@ -28,49 +28,35 @@
...
-import org.dswarm.persistence.util.DMPPersistenceUtil;
 
 public class SchemasResourceTest extends
 		BasicResourceTest<SchemasResourceTestUtils, SchemaServiceTestUtils, SchemaService, ProxySchema, Schema, Long> {
 
-	private AttributesResourceTestUtils		attributesResourceTestUtils;
+	private ClaszesResourceTestUtils claszesResourceTestUtils;
 
-	private ClaszesResourceTestUtils		claszesResourceTestUtils;
+	private SchemaAttributePathInstancesResourceTestUtils schemaAttributePathInstancesResourceTestUtils;
```
Comment (janpolowinski) [inline controller/src/test/java/org/dswarm/controller/resources/schema/test/SchemasResourceTest.java]: I'd suggest to make one more refaktoring step and search and replace "schemaAttributePathInstance_" (case sensitive search to avoid renaming the class) by "schemaAPInstance_" for at least a bit more readability. To be consistent we could also do this for "mappingAttributePathInstance". Especially when even longer names are constructed like schemaAttributePathInstanceAttributePathAttributes .... this may be critical. The meaning of AP can easily be derived in an IDE by looking at the type then.

File: controller/src/test/resources/project_to_remove_mapping_from_with_dummy_IDs.json [hunk 7]
```diff
@@ -1,331 +1,381 @@
...
-                "attributes": [{
-                    "name": "type",
-                    "uri": "http://www.w3.org/1999/02/22-rdf-syntax-ns#type",
-                    "id": -2323909237623843453
-                }]
-            }
+    "schema" : {
+      "name" : "atMostTwoRows.csv schema",
+      "id" : -4183243486324374530,
+      "attribute_paths" : [ {
```
Comment (janpolowinski) [inline controller/src/test/resources/project_to_remove_mapping_from_with_dummy_IDs.json]: I think it's intentional and fine to keep the name "attribute_paths" here and not use "attribute_path_instances" not to overcomplicate the JSON. However, at other places in the Java code, I'd vote for replacing attributePath by attributePathInstance when the type is an instance to avoid confusion. (I'll make an extra comment when I see it next time).

File: converter/src/test/java/org/dswarm/converter/adapt/ConverterModelTest.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * @author sbarthel
+ * @author tgaengler
+ */
+public class ConverterModelTest extends ModelTest {
+
+	private static final Logger log = LoggerFactory.getLogger(ConverterModelTest.class);
+
+	//@Test
```
Comment (janpolowinski) [inline converter/src/test/java/org/dswarm/converter/adapt/ConverterModelTest.java]: maybe also here some explanation on the purpose and usage of the disabled tests?

File: converter/src/test/java/org/dswarm/converter/adapt/JsonSchemaTransformer.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.base.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JsonSchemaTransformer {
```
Comment (janpolowinski) [inline converter/src/test/java/org/dswarm/converter/adapt/JsonSchemaTransformer.java]: some docu here and/or in the wiki?

File: persistence/src/main/java/org/dswarm/persistence/model/schema/AttributePathInstanceType.java [hunk 10]
```diff
@@ -19,7 +19,7 @@
  * The attribute path instance type enum. An attribute path instance type indicates the attribute path instance type of an
  * attribute path instance.<br>
  * 
- * @author tgaengler (created), Mar 18, 2013
+ * @author tgaengler (created), Mar 18, 2013 // TODO : ?
```
Comment (janpolowinski) [inline persistence/src/main/java/org/dswarm/persistence/model/schema/AttributePathInstanceType.java:22]: maybe do some clean up here

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/internalmodel/BiboDocumentSchemaBuilder.java [hunk 11]
```diff
@@ -33,48 +33,48 @@ public Schema buildSchema() {
...
 
 		// extra (added to have some details on creator/contributor resources):
-		tempSchema.addAttributePath(builder.parsePrefixPath("dcterms:creator/rdf:type"));
-		tempSchema.addAttributePath(builder.parsePrefixPath("dcterms:creator/foaf:familyName"));
-		tempSchema.addAttributePath(builder.parsePrefixPath("dcterms:creator/foaf:givenName"));
-		tempSchema.addAttributePath(builder.parsePrefixPath("dcterms:contributor/rdf:type"));
-		tempSchema.addAttributePath(builder.parsePrefixPath("dcterms:contributor/foaf:familyName"));
-		tempSchema.addAttributePath(builder.parsePrefixPath("dcterms:contributor/foaf:givenName"));
+		tempSchema.addAttributePath(builder.parseAsAttributePathInstance("dcterms:creator/rdf:type"));
+		tempSchema.addAttributePath(builder.parseAsAttributePathInstance("dcterms:creator/foaf:familyName"));
```
Comment (janpolowinski) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/internalmodel/BiboDocumentSchemaBuilder.java:73]: this is still the old schema building code. there should be something newer in my fork, we may want to merge soon. This shouldn't cause errors, but I think the front end does not yet need to be able to handle both - flat structures with long attribute paths AND the subschema-approach. Longer paths as given here could only be interpreted to create a node of type rdfs:Resource, since no other type information is given.

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java [hunk 12]
```diff
@@ -224,55 +252,208 @@
...
+	}
+
+	public AttributePath getDctermsTitleAP() throws Exception {
+
+		return createObject(AttributeServiceTestUtils.DCTERMS_TITLE);
+	}
+
+	public AttributePath getDctermsCreatorFoafFirstnameAP() throws Exception {
+
+		return createObject(DCTERMS_CREATOR__FOAF_FIRSTNAME_AP);
```
Comment (janpolowinski) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java:428]: Do we really need constants even for attribute paths? I'm not sure, whether reuse is worth this.

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java [hunk 13]
```diff
@@ -224,55 +252,208 @@
...
+	 */
+	@Override
+	public AttributePath createDefaultObject() throws Exception {
+
+		return getDctermsTitleDctermHaspartDctermsTitleAP();
+	}
+
+	public AttributePath getDctermsTitleDctermHaspartDctermsTitleAP() throws Exception {
+
+		return createObject(DCTERMS_TITLE__DCTERMS_HASPART__DCTERMS_TITLE_AP);
```
Comment (janpolowinski) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java:393]: Maybe it's sufficient to have either some methods with very long specific names OR have these very long constants. Now we have both!

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java [hunk 14]
```diff
@@ -390,6 +570,120 @@ public void deleteObject(final AttributePath object) {
...
+
+		return createAttributePath(attributes);
+	}
+
+	public AttributePath getNonCachedDctermsTitleDctermsHaspartDctermsTitleAP() throws Exception {
+
+		return getNonCachedAttributePath(commonAttributePathsMap.get(DCTERMS_TITLE__DCTERMS_HASPART__DCTERMS_TITLE_AP));
+	}
+
+	@Override
```
Comment (janpolowinski) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java:640]: please check

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributeServiceTestUtils.java [hunk 15]
```diff
@@ -17,17 +17,30 @@
...
 import org.dswarm.persistence.model.schema.proxy.ProxyAttribute;
+import org.dswarm.persistence.model.types.Tuple;
 import org.dswarm.persistence.service.schema.AttributeService;
 import org.dswarm.persistence.service.test.utils.AdvancedDMPJPAServiceTestUtils;
 
 public class AttributeServiceTestUtils extends AdvancedDMPJPAServiceTestUtils<AttributeService, ProxyAttribute, Attribute> {
 
-	public static final Set<String>	excludeAttributes		= Sets.newHashSet();
-	public static final Set<String>	excludeSubAttributes	= Sets.newHashSet();
+	public static final String DCTERMS_TITLE = "http://purl.org/dc/terms/title";
```
Comment (janpolowinski) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributeServiceTestUtils.java:31]: We can refaktor this later, but these constants may also be useful at other places and not only in the test code. Should we move them simply to Attribute?

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/internalmodel/BiboDocumentSchemaBuilder.java (hunk 11 above)
Comment (boferri) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/internalmodel/BiboDocumentSchemaBuilder.java:73]: hm, dunno. I thought everything re. the new sub schema alignments in our internal schema shoudl be fine now. At least this was my intention. I think that your commit re. the new internal schema should be part of this MR, or? (you may also have a look at the init_internal_schema.sql script)

Comment (boferri) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/internalmodel/BiboDocumentSchemaBuilder.java:73]: It looks like that something with the (commits of the) MR is wrong. https://github.com/zazi/dswarm/blob/sprint-24/dd-701/persistence/src/test/java/org/dswarm/persistence/service/schema/test/internalmodel/BiboDocumentSchemaBuilder.java shows also the outdated code. however, https://github.com/zazi/dswarm/blob/sprint-26/dd-734/persistence/src/test/java/org/dswarm/persistence/service/schema/test/internalmodel/BiboDocumentSchemaBuilder.java should include the actual version of it.

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java (hunk 12 above)
Comment (boferri) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java:428]: It's just for our test precast part tool box. so I think: don't worry. It's always handy to have a short cut for common APs.

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java (hunk 13 above)
Comment (boferri) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java:393]: yepp, but both of them are useful. in general, the amount of those methods and identifiers is rather small (< 10). so both approaches are applicable.

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java (hunk 14 above)
Comment (boferri) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java:640]: this method is just to for getting a fresh, non-cached AP. because the default method caches the created APs (and will return the cached AP, if requested twice ...).

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributeServiceTestUtils.java (hunk 15 above)
Comment (boferri) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributeServiceTestUtils.java:31]: no, there is no need for this yet. albeit it's a common attribute. our platform is not restricted or biased to any vocabulary or term.

File: persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java (hunk 14 above)
Comment (janpolowinski) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java:640]: I think you thought my comment was on getNonCachedDctermsTitleDctermsHaspartDctermsTitleAP()? It was on the method below returning null. Is the "return null" intended?

Comment (boferri) [inline persistence/src/test/java/org/dswarm/persistence/service/schema/test/utils/AttributePathServiceTestUtils.java:640]: yes, this is just a method stub of a method that is not implement in this concrete ServiceTestUtils impl.
it might be better, if an appropriate exception will be thrown (on the other side, it's just test code ... :) )

File: controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java (hunk 2 above)
Comment (janpolowinski) [inline controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java]: What is the benefit of using Collection here instead of List? Schema will return ordered APInstances, since it returns the values from a linked map impl., but making the order of the collection explicit may "prevent" (or better discourage) developers from ignoring this order in further transformations. If there is no strong reason against a List, I'd change this and also change the return type of getAttributePaths to List.

File: converter/src/test/java/org/dswarm/converter/schema/test/XMLSchemaParserTest.java [hunk 16]
```diff
@@ -35,6 +51,29 @@
...
+	public void prepare() throws Exception {
+		GuicedTest.tearDown();
+		GuicedTest.startUp();
+		initObjects();
+		maintainDBService.createTables();
+		maintainDBService.truncateTables();
+	}
+
+	@After
+	public void tearDown2() throws Exception {
```
Comment (janpolowinski) [inline converter/src/test/java/org/dswarm/converter/schema/test/XMLSchemaParserTest.java:66]: having had some trouble to find out about the differences of tearDown2 (seems not to exist anymore in recent branches), tearDown3 and tearDown, I suggest to use a more descriptiveName for tearDown3 like tearDownAfterMethod or at least rename to tearDown2

File: controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java (hunk 2 above)
Comment (boferri) [inline controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java]: Yep, however interfaces hide implementation details. There was a certain reason to change this to Collection, but I can't remember anymore. I would tend to say that even a List interface itself doesn't say anything about the order (/the semantics) of the elements.

Comment (janpolowinski) [inline controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java]: I'll take this list/collection issue into the list of changes to be done later, but the list interface exactly adds the ordered aspect to a collection. From the Docs: "An ordered collection (also known as a sequence). "

File: controller/src/docs/ui (hunk 1 above)
Comment (janpolowinski) [inline controller/src/docs/ui:1]: @jnauber or @chrode - may be you can clarify if we need to undo this commit not to harm the sub-module structure? Sorry, I'm still not sure about the sub-module topic. We have a commit that whants to delete "controller/src/docs/ui" in the current pull request and wonder if this may have been commited unintentionally.

File: controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java (hunk 2 above)
Comment (boferri) [inline controller/src/main/java/org/dswarm/controller/resources/schema/SchemasResource.java]: afaik, it's just the issue that Map#values() delivers a Collection instead of a List (and were are utilising a Map<String, SAPI> right now in Schema for holding the SAPIs of the schema).
